### PR TITLE
feat(ui): Introduce experimental walletconnect

### LINKED ISSUE #940: feat(ui): Integrate eth wallet using walletconnect
As a replacement of the Radicle Registry, we want to integrate an Eth wallet in our application using `walletconnect`.

### TODO

- [x] Re-enable the wallet section
- [ ] Rename 'Wallet" to "Funding"
- [ ] Handle the connection to a wallet
   - [x] Connect
     - [ ] Timeout after X seconds with relevant error message 
   - [x] Disconnect
- When a wallet is connected
   - [x] React to user changing accounts in their wallet app
   - [ ] Handle disconnections
   - [ ] Handle wallet with 0 accounts

### PR DESCRIPTION
**Pull request by NunoAlexandre**
Closes #940 

In this branch, we introduce an experimental feature where we use  `walletconnect` to connect an Eth wallet to the Upstream. This experiment lives under the `Wallet` section for now and allows us to `connect` by scanning a QR Code using a compatible wallet. We also introduce a dependency on the Radicle Contracts and a script to deploy them to a local blockchain instance.

When building/running, we see js warnings which I don't yet know how to solve which does _not_ get in the way.
I am also seeing an issue with disconnecting from a connected wallet, where it seems that the underlying components don't quite expect to disconnect and a consecutive `connecti` attempt does nothing. This is ok for now and can be walked around by switching tabs, which creates a fresh wallet.

To give it a try, go to `Wallet` > `Connect`, use your preferred wallet and scan the QR Code the app shows. Once you approve the connection on your wallet, you should see your balance and address in the app. Try switching accounts from your wallet app to see how it reflects in the app.

### CODE REVIEW

File: package.json
Comment (cloudhead) [inline package.json:85]: Can we avoid using `ethers` since we are pulling in the web3.js provider for walletconnect?

Comment (NunoAlexandre) [reply package.json]: done!

**Review by rudolfs — COMMENTED** [review]
As discussed with @NunoAlexandre on IRC: I would prefer if we leave this in a feature branch until we have a clearer plan on how the functionality will look like and it is a bit more polished.

I fear that the added clutter of warnings might prevent us from detecting actual problems in future pull-requests if this gets merged.

```
(!) Circular dependencies
node_modules/hash-base/node_modules/readable-stream/lib/_stream_readable.js -> node_modules/hash-base/node_modules/readable-stream/lib/_stream_duplex.js -> node_modules/hash-base/node_modules/readable-stream/lib/_stream_readable.js
node_modules/hash-base/node_modules/readable-stream/lib/_stream_duplex.js -> node_modules/hash-base/node_modules/readable-stream/lib/_stream_writable.js -> node_modules/hash-base/node_modules/readable-stream/lib/_stream_duplex.js
node_modules/hash-base/node_modules/readable-stream/lib/_stream_duplex.js -> node_modules/hash-base/node_modules/readable-stream/lib/_stream_writable.js -> /Users/rudolfs/work/radicle-upstream-2/node_modules/hash-base/node_modules/readable-stream/lib/_stream_duplex.js?commonjs-proxy -> node_modules/hash-base/node_modules/readable-stream/lib/_stream_duplex.js
...and 13 more
(!) `this` has been rewritten to `undefined`
https://rollupjs.org/guide/en/#error-this-is-undefined
node_modules/@ethersproject/properties/lib.esm/index.js
1: "use strict";
2: var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
                    ^
3:     function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
4:     return new (P || (P = Promise))(function (resolve, reject) {
...and 1 other occurrence
(!) Plugin inject: rollup-plugin-inject: failed to parse /Users/rudolfs/work/radicle-upstream-2/node_modules/eccrypto-js/node_modules/secp256k1/lib/messages.json. Consider restricting the plugin to particular files via options.include
node_modules/eccrypto-js/node_modules/secp256k1/lib/messages.json
```

File: package.json
Comment (NunoAlexandre) [reply package.json:157]: Not for now, but this is a stable location for the period we are concerned with. We can’t point to the GitHub project directly (i.e. yarn add https://github.com/radicle-dev/radicle-contracts.git) because we need to create build artefacts (compile TS and the contracts) and yarn doesn't do this for git dependencies.

File: scripts/deploy-dev-contracts.js
Comment (NunoAlexandre) [reply scripts/deploy-dev-contracts.js:1]: That is a good idea but not very useful yet. I'd suggest we do that once we go on actually wire the contracts up in the upstream, as then we will have a better understanding of the needs and requirements for this bit.

File: scripts/deploy-dev-contracts.js
Comment (NunoAlexandre) [reply scripts/deploy-dev-contracts.js:13]: I think that in the future this should point to the mainnet instance.

File: ui/src/wallet.ts
Comment (NunoAlexandre) [reply ui/src/wallet.ts:44]: yap.

**Comment by NunoAlexandre** [comment]
> As discussed with @NunoAlexandre on IRC: I would prefer if we leave this in a feature branch until we have a clearer plan on how the functionality will look like and it is a bit more polished.
> 
> I fear that the added clutter of warnings might prevent us from detecting actual problems in future pull-requests if this gets merged.
> 
> ```
> (!) Circular dependencies
> node_modules/hash-base/node_modules/readable-stream/lib/_stream_readable.js -> node_modules/hash-base/node_modules/readable-stream/lib/_stream_duplex.js -> node_modules/hash-base/node_modules/readable-stream/lib/_stream_readable.js
> node_modules/hash-base/node_modules/readable-stream/lib/_stream_duplex.js -> node_modules/hash-base/node_modules/readable-stream/lib/_stream_writable.js -> node_modules/hash-base/node_modules/readable-stream/lib/_stream_duplex.js
> node_modules/hash-base/node_modules/readable-stream/lib/_stream_duplex.js -> node_modules/hash-base/node_modules/readable-stream/lib/_stream_writable.js -> /Users/rudolfs/work/radicle-upstream-2/node_modules/hash-base/node_modules/readable-stream/lib/_stream_duplex.js?commonjs-proxy -> node_modules/hash-base/node_modules/readable-stream/lib/_stream_duplex.js
> ...and 13 more
> (!) `this` has been rewritten to `undefined`
> https://rollupjs.org/guide/en/#error-this-is-undefined
> node_modules/@ethersproject/properties/lib.esm/index.js
> 1: "use strict";
> 2: var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
>                     ^
> 3:     function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
> 4:     return new (P || (P = Promise))(function (resolve, reject) {
> ...and 1 other occurrence
> (!) Plugin inject: rollup-plugin-inject: failed to parse /Users/rudolfs/work/radicle-upstream-2/node_modules/eccrypto-js/node_modules/secp256k1/lib/messages.json. Consider restricting the plugin to particular files via options.include
> node_modules/eccrypto-js/node_modules/secp256k1/lib/messages.json
> ```

Yes let's do that. Created `feat/funding`, to which this PR is now targeted.

File: ui/Screen/Profile/Wallet.svelte
Comment (sarahscott) [inline ui/Screen/Profile/Wallet.svelte:1]: Let's type this one since it's easy 😊

File: ui/src/wallet.ts
Comment (sarahscott) [inline ui/src/wallet.ts:50]: Why is this an `any`?

File: ui/src/wallet.ts
Comment (sarahscott) [inline ui/src/wallet.ts:54]: If there's no `accountAddress`, should we call `getBalance` at all?

File: ui/src/wallet.ts
Comment (sarahscott) [inline ui/src/wallet.ts:55]: I assume this might be captured in the `handle error` TODO, but it'd be nice to verify this connection before updating the `status`.

File: ui/src/wallet.ts
Comment (sarahscott) [inline ui/src/wallet.ts:82]:
```ts
stateStore.set({ status: Status.NotConnected });
```

File: ui/src/wallet.ts
Comment (sarahscott) [inline ui/src/wallet.ts:99]: Shouldn't the `status` be updated to `Disconnected` here as well?

**Comment by sarahscott** [comment]
Looks awesome so far 💯 ! In the interest of keeping things moving and not letting this feature branch diverge from master for too long, maybe we can aim to merge this in once the console error concerns are figured out.

File: package.json
Comment (cloudhead) [reply package.json:157]: We should definitely publish an npm package though soon.

**Comment by rudolfs** [comment]
> maybe we can aim to merge this in once the console error concerns are figured out

That would be fine by me.

File: ui/src/wallet.ts
Comment (NunoAlexandre) [reply ui/src/wallet.ts:34]: I see doing so being useful if the config is environment-dependent but simply moving these hardcoded values to a different file while they are used in a single place doesn't seem to add much value. Could you explain what benefits you see?

Comment (rudolfs) [reply ui/src/wallet.ts:34]: I think it would be easier to find if all the configuration knobs have meaningful names and in one place, so I don't have to hunt around unfamiliar code.

But yeah, as this is just an experiment, perhaps it's too early for that.

File: ui/src/wallet.ts
Comment (xla) [inline ui/src/wallet.ts:15]: There is a lot of connecting in this pr. :laughing: It would be helpful to give at leas this field and type a different name.

File: ui/src/wallet.ts
Comment (xla) [reply ui/src/wallet.ts:34]: Will there ever be a different walletconnect endpoint to be used?

File: ui/src/wallet.ts
Comment (xla) [inline ui/src/wallet.ts:31]: Is this constructor adding much value? It's called once in the wallet component and otherwise not used in other ways. I think removing it and just export the relevant parts from the module directly is as ergonomic and the more widely used style in the rest of the code-base.

**Review by xla — CHANGES_REQUESTED** [review]
Extra dope!

take my money

File: ui/src/wallet.ts
Comment (NunoAlexandre) [reply ui/src/wallet.ts:34]: don't think so. that's also the value that walletconnect uses in their sample project and haven't seen another alternative.

File: package.json
Comment (NunoAlexandre) [reply package.json:157]: Ah! Does that mean that if we had a `.npmignore` file to the `Radicle Contracts` that pointing to the git project should work?

Comment (xla) [reply package.json:157]: AFAIU yes as the issue seems to be that yarn cleans up the build artefacts.

Comment (NunoAlexandre) [reply package.json:157]: will give it a shot

File: ui/src/wallet.ts
Comment (NunoAlexandre) [reply ui/src/wallet.ts]: this had been written by Thomas. `provider.enable()` returns `Promise<unknown>`, I assume the `as any` was the quickest workout. Changed it to `as string[]`.

File: ui/src/wallet.ts
Comment (NunoAlexandre) [reply ui/src/wallet.ts]: the error is now handled :+1: 
as far as I have seen, if  `provider.enable()` succeeds, the connection is established, so we are good.

File: ui/src/wallet.ts
Comment (NunoAlexandre) [reply ui/src/wallet.ts:103]: This is called as a result of calling `disconnect()` or the provider disconnecting, so we shouldn't need to do it here. However, it might be the case that it's the direct `connector` connecting going off, so we better cover it. Thanks!

File: rollup.app.js
Comment (NunoAlexandre) [reply rollup.app.js:46]: This was added by Thomas fyi, and I believe that yes, it's necessary by some of the new dependencies. Will double-check.

File: tsconfig.json
Comment (NunoAlexandre) [reply tsconfig.json:6]: Same as https://github.com/radicle-dev/radicle-upstream/pull/952#discussion_r494151273

File: ui/src/wallet.ts
Comment (NunoAlexandre) [reply ui/src/wallet.ts:15]: do you have a suggestion? thought the same but name it `Wallet` doesn't help and could come up with a better name.

File: ui/src/wallet.ts
Comment (NunoAlexandre) [reply ui/src/wallet.ts:31]: If we don't have it as a "constructor" we can't build a new instance as needed, which now is necessary since we face issues when disconnecting, were creating a new instance is a useful workaround for now.

Comment (NunoAlexandre) [reply ui/src/wallet.ts:31]: Btw, in the future, we want to offer more alternatives to WalletConnect to I will move this to a `walletconnect.ts` module and have this constructor receive a "ConnectorOption` or so and call the appropriate constructor.

File: package.json
Comment (NunoAlexandre) [reply package.json:85]: We actually need ethers in ./scripts/deploy-dev-contracts.js, added it back. it's a dev dependency btw.

File: package.json
Comment (NunoAlexandre) [reply package.json:158]: gave it a try but it doesn't work. the issue being that [Yarn does not run `prepack` when installing Git dependencies](https://github.com/yarnpkg/yarn/issues/5047)

File: ui/src/wallet.ts
Comment (NunoAlexandre) [reply ui/src/wallet.ts]: fixed :ok_hand:

File: scripts/deploy-dev-contracts.js
Comment (NunoAlexandre) [reply scripts/deploy-dev-contracts.js:1]: Well, just added it, it will be useful anyway. Thanks for the suggestion, @rudolfs!

File: scripts/deploy-dev-contracts.js
Comment (NunoAlexandre) [reply scripts/deploy-dev-contracts.js:13]: Added a yarn command as suggested by @rudolfs and a corresponding (minimal) documentation on `DEVELOPMENT.md`. I'd prefer to not be more extensive now or else we might be documenting stuff that will change soon enough.

**Comment by rudolfs** [comment]
I tried to connect my mobile Metamask wallet. The connection seemed to work, but I don't see my balance, do I need to run some daemon locally (if so, how do I set that up)?

Screenshot 2020-09-24 at 12 59 51
Screenshot 2020-09-24 at 13 00 08

**Comment by NunoAlexandre** [comment]
> I tried to connect my mobile Metamask wallet. The connection seemed to work, but I don't see my balance, do I need to run some daemon locally (if so, how do I set that up)?
> 
> Screenshot 2020-09-24 at 12 59 51
> 
> Screenshot 2020-09-24 at 13 00 08

ah, that is right, sorry! This should answer all your getting started needs: https://www.notion.so/monadic/Spike-Walletconnect-on-Upstream-18a27328098845b091f25bf0b429a375

**Comment by NunoAlexandre** [comment]
Superseded by #974 

@geigerzaehler leave it up to you to close it since there are open discussions that'd be better addressed by you :+1:

**Pull request closed without merging**
